qemu/tests/balloon_memhp.py

```diff
@@ -0,0 +1,88 @@
+from virttest import utils_numeric
+from virttest.utils_test.qemu import MemoryHotplugTest
+from qemu.tests.balloon_check import BallooningTestWin
+from qemu.tests.balloon_check import BallooningTestLinux
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Balloon and memory hotplug test, i.e. guest-stats-polling-interval test
+    1) boot a guest with balloon device
+    2) enable and check driver verifier in guest(only for windows guest)
+    3) install balloon service in guest(only for windows guest)
+    4) evict balloon
+    5) hotplug memory to guest
+    6) check balloon and guest memory
+    7) enlarge balloon
+    8) unplug the memory in step 5 (only for linux guest)
+    9) check balloon and guest memory
+    10) uninstall balloon service and clear driver verifier(only for
+       windows guest)
+    """
+    def check_memory(op="hotplug"):
+        """
+        Check guest memory after doing operation op
+        :param op: the operation done to guest before
+        """
+        if params['os_type'] == 'windows':
+            memhp_test.check_memory(vm)
+        else:
+            if op == "hotplug":
+                expected_mem = new_mem + mem_dev_sz
+            else:
+                expected_mem = new_mem - mem_dev_sz
+            guest_mem_size = memhp_test.get_guest_total_mem(vm)
+            threshold = float(params.get("mem_threshold", 400))
+            if expected_mem - guest_mem_size > threshold:
+                msg = ("Assigned '%s MB' memory to '%s'"
+                       "but, '%s MB' memory detect by OS" %
+                       (expected_mem, vm.name, guest_mem_size))
+                test.fail(msg)
+
+    error_context.context("Boot guest with balloon device", logging.info)
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    session = vm.wait_for_login()
+    if params['os_type'] == 'windows':
+        driver_name = params.get("driver_name", "balloon")
+        session = utils_test.qemu.windrv_check_running_verifier(session, vm,
+                                                                test, driver_name)
+        balloon_test = BallooningTestWin(test, params, env)
+        error_context.context("Config balloon service in guest",
+                              logging.info)
+        balloon_test.configure_balloon_service(session)
+    else:
+        balloon_test = BallooningTestLinux(test, params, env)
+
+    memhp_test = MemoryHotplugTest(test, params, env)
+
+    mem_dev_sz = params["size_mem"]
+    mem_dev_sz = int(utils_numeric.normalize_data_size(mem_dev_sz, "M"))
+    target_mem = params["target_mem"]
+
+    try:
+        min_sz, max_sz = balloon_test.get_memory_boundary()
+        new_mem = int(random.uniform(min_sz, max_sz))
+        balloon_test.balloon_memory(new_mem)
+        memhp_test.hotplug_memory(vm, target_mem)
+        check_memory(op="hotplug")
+        new_mem = int(random.uniform(min_sz + mem_dev_sz, max_sz + mem_dev_sz))
```

Comment:
I think it's better to balloon guest mem to maximal(ori_mem+mem_dev_sz) after plug to make sure this part of mem can be used by balloon.

Reply:
And after that you can balloon to another random value to check evict also works.
btw, the minimal mem may change if guest total mem changes, better to get min_sz and max_sz again after plug.